added minimal_scheduling_real_time_tutorial.cpp

JanStaschulat opened the pull request:
example with two subscriptions, two threads, two executors

On subscription is executed in a thread with real-time prioriy. The other subscription is exeucted in a thread with default priority.

JanStaschulat committed:
added minimal_scheduling_real_time_tutorial.cpp

Signed-off-by: Jan Staschulat <jan.staschulat@de.bosch.com>

JanStaschulat committed:
updated README

Signed-off-by: Jan Staschulat <jan.staschulat@de.bosch.com>

JanStaschulat commented:
@carlossvg what do you think about this code example and the tutorial?

https://github.com/ros-realtime/ros2_documentation/blob/feature/update-real-time/source/Tutorials/Advanced/Real-Time/Real-Time-Processing-with-multiple-threads.rst

By the way, I do not have write access permissions on this repo. Therefore I had to create a fork on my github-account and create this PR.

carlossvg commented on minimal_scheduling/minimal_scheduling_real_time_tutorial.cpp:
```diff
@@ -0,0 +1,153 @@
+  // middleware threads are not configured with real-time priority
+  // because scheduling configuration of main thread is not real-time.
+
+  // Note: you can only add complete nodes to an Executor. That is:
```
@JanStaschulat It is possible to only a subset of subscriptions to an executor if they are excluded from the default callback group. There is an example here: https://github.com/ros-realtime/ros2-realtime-examples/blob/c09aa1f26b249b21366288f76986d7fd13e7a808/minimal_scheduling/minimal_scheduling_callback_group.cpp

I'm fine showing mixed criticality using both options:
- Using callback groups
- Using separate nodes

carlossvg replied:
>@carlossvg what do you think about this code example and the tutorial?

Looks good to me. I would remove or refine the "Resources" section. Most of the links don't make sense for a tutorial IMO.

WRT the asciinema, we should make sure it is okay to use them as part of the ROS 2 docs since it relies on a third party platform and on a specific user account (mine in this case). 

> By the way, I do not have write access permissions on this repo. Therefore I had to create a fork on my github-account and create this PR.

I will give to writer permissions.

JanStaschulat replied:
Yes, mixed-criticality can be achieved with different functionality. With this example I wanted to show the case with only two threads and two executors. 

In a second tutorial, we can describe mixed criticality with your existing example with callback groups. I thought, this is a bit more advanced, so I wanted to have a simpler code example as first option.

JanStaschulat commented:
@carlossvg What do you mean with ...

>  I would remove or refine the "Resources" section. Most of the links don't make sense for a tutorial IMO.

Which source are you referring to? Some text in the new tutorial? I am working in the branch [feature/update-realtime](https://github.com/ros-realtime/ros2_documentation/tree/feature/update-real-time/source/Tutorials/Advanced/Real-Time) and added one file: 
https://github.com/ros-realtime/ros2_documentation/tree/feature/update-real-time/source/Tutorials/Advanced/Real-Time

carlossvg commented:
> @carlossvg What do you mean with ...
> 
> > I would remove or refine the "Resources" section. Most of the links don't make sense for a tutorial IMO.
> 
> Which source are you referring to? Some text in the new tutorial? I am working in the branch [feature/update-realtime](https://github.com/ros-realtime/ros2_documentation/tree/feature/update-real-time/source/Tutorials/Advanced/Real-Time) and added one file: https://github.com/ros-realtime/ros2_documentation/tree/feature/update-real-time/source/Tutorials/Advanced/Real-Time

@JanStaschulat I mean this section: https://github.com/ros-realtime/ros2-realtime-examples/blob/c09aa1f26b249b21366288f76986d7fd13e7a808/minimal_scheduling/README.md#resources

JanStaschulat commented:
okay, understood. 

Question regarding the tutorial, I explained in the tutorial the entire source code with thread joining etc. Do you think that a tutorial should be complete or should it only describe the essentials, like thread creation and Executor management?

JanStaschulat committed:
updated ressources

Signed-off-by: Jan Staschulat <jan.staschulat@de.bosch.com>

JanStaschulat commented:
Hi @carlossvg, I updated the ressource section in README. 
- structured the list a bit
- added two links to docs.ros.org

Are there any other open issues before this PR can be merged?

carlossvg commented on minimal_scheduling/CMakeLists.txt:
```diff
@@ -26,11 +26,15 @@ ament_target_dependencies(minimal_scheduling_spin_thread rclcpp std_msgs)
 add_executable(minimal_scheduling_callback_group minimal_scheduling_callback_group.cpp)
 ament_target_dependencies(minimal_scheduling_callback_group rclcpp std_msgs)
 
+add_executable(minimal_scheduling_real_time_tutorial minimal_scheduling_real_time_tutorial.cpp)
```
@JanStaschulat This executable might be used independently of the ROS 2 tutorial. I would change the name of the executable to something describing what this example is doing instead of describing where it is used.

JanStaschulat replied:
changed filename to `minimal_scheduling_real_time_executor`

carlossvg commented:
@JanStaschulat I left one comment about the executable name. Other than that looks good to me.

JanStaschulat committed:
changed filename

Signed-off-by: Jan Staschulat <jan.staschulat@de.bosch.com>

JanStaschulat commented:
@carlossvg changed the filename, so from my side the PR could be merged.

carlossvg commented:
> @carlossvg changed the filename, so from my side the PR could be merged.

Thanks, yes, let's merge it!
